system_tests/program_ink_test.go

```diff
@@ -0,0 +1,576 @@
+			hostio:      "chainid",
+			expectedInk: HOSTIO_INK,
+		},
+		{
+			hostio:      "block_coinbase",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+		{
+			hostio:      "block_gas_limit",
+			expectedInk: HOSTIO_INK,
+		},
+		{
+			hostio:      "block_number",
+			expectedInk: HOSTIO_INK,
+		},
+		{
+			hostio:      "block_timestamp",
+			expectedInk: HOSTIO_INK,
+		},
+		{
+			hostio:      "contract_address",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+		{
+			hostio:      "math_div",
+			args:        []any{big.NewInt(1), big.NewInt(3)},
+			expectedInk: 43520,
+		},
+		{
+			hostio:      "math_mod",
+			args:        []any{big.NewInt(1), big.NewInt(3)},
+			expectedInk: 43520,
+		},
+		{
+			hostio:      "math_add_mod",
+			args:        []any{big.NewInt(1), big.NewInt(3), big.NewInt(5)},
+			expectedInk: 49560,
+		},
+		{
+			hostio:      "math_mul_mod",
+			args:        []any{big.NewInt(1), big.NewInt(3), big.NewInt(5)},
+			expectedInk: 52660,
+		},
+		{
+			hostio:      "msg_sender",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+		{
+			hostio:      "msg_value",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+		{
+			hostio:      "tx_gas_price",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+		{
+			hostio:      "tx_ink_price",
+			expectedInk: HOSTIO_INK,
+		},
+		{
+			hostio:      "tx_origin",
+			expectedInk: HOSTIO_INK + PTR_INK,
+		},
+	} {
+		t.Run(tc.hostio, func(t *testing.T) {
+			solFunc := matchSnake.ReplaceAllStringFunc(tc.hostio, func(s string) string {
+				return strings.ToUpper(strings.TrimPrefix(s, "_"))
+			})
+			data := encodeHostioTestCalldata(t, solFunc, tc.args)
+			checkInkUsage(t, builder, stylusProgram, tc.hostio, tc.hostio, data, nil, tc.expectedInk)
+		})
+	}
+}
+
+func TestPowInkUsage(t *testing.T) {
+	t.Parallel()
+
+	builder := setupGasCostTest(t)
+	auth := builder.L2Info.GetDefaultTransactOpts("Owner", builder.ctx)
+	stylusProgram := deployWasm(t, builder.ctx, auth, builder.L2.Client, rustFile("hostio-test"))
+
+	hostio := "math_pow"
+
+	for _, tc := range []struct {
+		exponentNumBytes uint
+		expectedInk      uint64
+	}{
+		{exponentNumBytes: 1, expectedInk: 61520},
+		{exponentNumBytes: 2, expectedInk: 79020},
+		{exponentNumBytes: 10, expectedInk: 219020},
+		{exponentNumBytes: 32, expectedInk: 604020},
+	} {
+		name := fmt.Sprintf("%v%v", hostio, tc.exponentNumBytes)
+		t.Run(name, func(t *testing.T) {
+			exponent := new(big.Int).Lsh(big.NewInt(1), tc.exponentNumBytes*8-1)
+			args := []any{big.NewInt(1), exponent}
+			data := encodeHostioTestCalldata(t, "mathPow", args)
+			checkInkUsage(t, builder, stylusProgram, hostio, name, data, nil, tc.expectedInk)
+		})
+	}
+}
+
+func TestStorageInkCost(t *testing.T) {
+	t.Parallel()
+
+	builder := setupGasCostTest(t)
+	auth := builder.L2Info.GetDefaultTransactOpts("Owner", builder.ctx)
+	stylusProgram := deployWasm(t, builder.ctx, auth, builder.L2.Client, rustFile("multicall"))
+
+	store_hostio := "storage_flush_cache"
+	load_hostio := "storage_load_bytes32"
```

Comment:
can we have a test that loads the same storage location twice?
it should cost less the second time (cold vs hot storage).
also interesting to see two loads vs. two loads with a cache flush between them.